From 6fd4b7d75f413d80bf6360dffa60a04dc10cc8e1 Mon Sep 17 00:00:00 2001
From: "Dharageswari.R" <dharageswari.r@intel.com>
Date: Tue, 23 Jun 2015 16:07:58 +0530
Subject: [PATCH] ASoC: core: Add Fix-up for BackEnd dai-link

In-order to support the DSP's like Intel which needs
dynamic configuration of CODEC<->CODEC dai_link params.

For such instances, a fix-up method is provided to the machine
driver to (re)configure the DAI link params based on the dynamic
selection of backend DAI configuration.

Change-Id: I5b58aff225e279d0217e1264b2f5d235f066fc35
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-12531
Signed-off-by: Dharageswari.R <dharageswari.r@intel.com>
Signed-off-by: Samreen Nilofer <samreen.nilofer@intel.com>
---
 include/sound/soc.h  |  3 +++
 sound/soc/soc-dapm.c | 18 ++++++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/include/sound/soc.h b/include/sound/soc.h
index 77d0c0fcf26a..3b9bed1c7705 100644
--- a/include/sound/soc.h
+++ b/include/sound/soc.h
@@ -689,6 +689,9 @@ struct snd_soc_dai_link {
 	int (*be_hw_params_fixup)(struct snd_soc_pcm_runtime *rtd,
 			struct snd_pcm_hw_params *params);
 
+	/* BE fixup method for CODEC<->CODEC dailink */
+	int (*be_fixup)(struct snd_soc_dai_link *dai_link,
+			struct snd_soc_dai *dai);
 	/* machine stream operations */
 	const struct snd_soc_ops *ops;
 	const struct snd_soc_compr_ops *compr_ops;
diff --git a/sound/soc/soc-dapm.c b/sound/soc/soc-dapm.c
index b407af7c8920..276c793d9aba 100644
--- a/sound/soc/soc-dapm.c
+++ b/sound/soc/soc-dapm.c
@@ -3530,6 +3530,15 @@ static int snd_soc_dai_link_event(struct snd_soc_dapm_widget *w,
 
 		if (source->driver->ops && source->driver->ops->hw_params) {
 			substream.stream = SNDRV_PCM_STREAM_CAPTURE;
+
+			if (w->dai_link->be_fixup) {
+				ret =  w->dai_link->be_fixup(w->dai_link, source);
+
+				if (ret != 0) {
+					dev_err(source->dev, "ASoC: fix_up_be()	failed: %d\n", ret);
+					goto out;
+				}
+			}
 			ret = source->driver->ops->hw_params(&substream,
 							     params, source);
 			if (ret != 0) {
@@ -3541,6 +3550,15 @@ static int snd_soc_dai_link_event(struct snd_soc_dapm_widget *w,
 
 		if (sink->driver->ops && sink->driver->ops->hw_params) {
 			substream.stream = SNDRV_PCM_STREAM_PLAYBACK;
+
+			if (w->dai_link->be_fixup) {
+				ret =  w->dai_link->be_fixup(w->dai_link, sink);
+
+				if (ret != 0) {
+					dev_err(source->dev, "ASoC: fix_up_be() failed: %d\n", ret);
+					goto out;
+				}
+			}
 			ret = sink->driver->ops->hw_params(&substream, params,
 							   sink);
 			if (ret != 0) {
-- 
1.9.1

